File: Graph/Dijstra/743. Network Delay Time/main.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

/*
	Input: times = [[2,1,1],[2,3,1],[3,4,1]], n = 4, k = 2
	Output: 2
*/
func main() {
	fmt.Println(networkDelayTime([][]int{{2,1,1}, {2,3,1},{3,4,1}}, 4, 2))
}

type node struct {
	key  int
	cost int
}

func networkDelayTime(times [][]int, n int, k int) int {
	adj := make(map[int][]node) // construct graph
	for _, time := range times {
		src := time[0]
		dest := time[1]
		cost := time[2]
		adj[src] = append(adj[src], node{key: dest, cost: cost})
	}

	timeToReach := make([]int, n+1)
	for i := 0; i <= n; i++ {
		timeToReach[i] = math.MaxInt32
	}
	timeToReach[k] = 0 // start point

	h := &minHeap{}
	heap.Init(h)
	heap.Push(h, node{k, 0})

	for len(*h) != 0 {
		currNode := heap.Pop(h).(node)
		currKey := currNode.key
		currCost := currNode.cost

		if currCost > timeToReach[currKey] { continue }
		for _, nextNode := range adj[currKey]{
			nextKey := nextNode.key
			nextCost := nextNode.cost
			if currCost + nextCost < timeToReach[nextKey]{
				timeToReach[nextKey] = currCost + nextCost
				heap.Push(h, node{nextKey, currCost + nextCost})
			}
		}
	}

	maxTime := 0
	for i := 1; i <= n; i++ {
		if maxTime < timeToReach[i]{
			maxTime = timeToReach[i]
		}
	}

	if maxTime == math.MaxInt32{
		return -1
	}

	return maxTime
}

type minHeap []node

func (m minHeap) Len() int {
	return len(m)
}

func (m minHeap) Swap(i, j int) {
	m[i], m[j] = m[j], m[i]
}

func (m minHeap) Less(i, j int) bool {
	return m[i].cost < m[j].cost
}

func (m *minHeap) Push(x interface{}) {
	*m = append(*m, x.(node))
}

func (m *minHeap) Pop() interface{} {
	n := len(*m)
	p := (*m)[n-1]
	*m = (*m)[:n-1]
	return p
}

File: Graph/Dijstra/743. Network Delay Time/main.py
import sys
from functools import total_ordering
from queue import PriorityQueue
from typing import List


@total_ordering
class Node:
    def __init__(self, num, cost):
        self.num = num
        self.cost = cost

    def __lt__(self, other):
        return self.cost < other.cost

    def __eq__(self, other):
        return self.cost == other.cost

    def __repr__(self):
        return "num = {} and cost = {}".format(self.num, self.cost)


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # build adj graph
        adj = {}
        for time in times:
            start = time[0]
            end = time[1]
            cost = time[2]
            if start not in adj:
                adj[start] = []
            adj[start].append(Node(end, cost))

        time_to_reach = [sys.maxsize] * (n + 1)
        time_to_reach[k] = 0

        queue = PriorityQueue()
        queue.put(Node(k, 0))

        while queue.qsize() != 0:
            node = queue.get()
            if node.cost > time_to_reach[node.num]:
                continue
            if adj.get(node.num) is None:
                continue
            for nei in adj.get(node.num):
                if nei.cost + node.cost < time_to_reach[nei.num]:
                    time_to_reach[nei.num] = nei.cost + node.cost
                    queue.put(Node(nei.num, nei.cost + node.cost))

        # start from 1 find the max time
        max_time = -1
        for i in range(1, len(time_to_reach)):
            if time_to_reach[i] > max_time:
                max_time = time_to_reach[i]

        if max_time == sys.maxsize:
            return -1
        else:
            return max_time